### PR TITLE
Aztec: Implements basic state machine for editor state

### LINKED ISSUE #6422: Aztec: Refactor Publish Button logic into a state machine
Pull out all of the logic for what the publish button says/does and put into a reusable container that can be shared among the three different editors.

### PR DESCRIPTION
**Pull request by astralbodies**
Closes #6422 

This adds the Publish button (previously called Post) to the Aztec editor. Instead of using a series of hacky methods to determine the state of the button it extracts the logic into clearly defined units of state that is unit testable. Other variables that affect state of the editor are also in the state context.

This is partially based off of @frosty's PostEditorSaveAction enum and WPPostViewController extension but made to be more abstract.

The main things this solves are:
* Publish button title.
* Publish button state (enabled/disabled).
* What text is shown during the publish process (Publishing...).

What it can/should solve but isn't part of this PR:
* Is Post-Post shown?
* Is there a secondary publish button and what does it say? (publish now, save draft, etc)
* Anything else that's related to the state of the editor that shouldn't turn into a pile of crazy If statements.

Some other TODOs that aren't part of this PR:
* **Submit for Review** is a lot of text to fit on the button label. Maybe just Submit? How are we handling it in cases where the button text is just too long in other languages?
* Submit for Review is also not currently supported in the hybrid editor. May need to validate how this all works. Currently the Aztec editor supports it but we're not setting the Bool properly if the user has or doesn't have the permission to publish.
* Update the state context when the post is being published to disable the publish button.
* Add some more unit tests to make sure all state switches are accounted for.

## How to Test
Make sure the publish button shows the right text and is enabled in these scenarios:
1. New post, no content. (Publish, disabled)
2. New post, with content. (Publish, enabled)
3. New post, set to draft. (Save)
4. New post, future dated. (Schedule)
5. New post, past dated. (Publish)
6. Existing published post. (Update)
7. Existing scheduled post. (Update)
8. Existing draft post. (Save)

There are a bunch of other scenarios but you get the point.

Needs review: @frosty, @jleandroperez

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift [hunk 1]
```diff
@@ -150,18 +156,46 @@ class AztecPostViewController: UIViewController {
...
+    fileprivate(set) lazy var postEditorStateContext: PostEditorStateContext = {
+        let originalPostStatus: PostStatus
+
+        if let originalPost = self.post.original,
+            let postStatus = originalPost.status,
+            originalPost.hasRemote() {
+            originalPostStatus = PostStatus(rawValue: postStatus) ?? .draft
+        } else {
+            originalPostStatus = .draft
+        }
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift]: Nipticky // feel free to disregard!!

Potential candidate for a helper method:

```
enum PostStatus {
   static func fromPost(post: Post) -> PostStatus {
 //...
   }
}
```

File: WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift [hunk 2]
```diff
@@ -150,18 +156,46 @@ class AztecPostViewController: UIViewController {
...
+    }()
+
     // MARK: - Lifecycle Methods
 
     init(post: AbstractPost) {
         self.post = post
+
         super.init(nibName: nil, bundle: nil)
+
+        addObservers(toPost: post)
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift]: I _think_ it's safe to remove this `addObservers`, since it's already being done in the **didSet**.

Unless i'm missing something!

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+            if hasContent != oldValue {
+                publishActionAllowed = hasContent && !isBeingPublished
+            }
+        }
+    }
+
+    fileprivate var isBeingPublished = false {
+        didSet {
+            if isBeingPublished != oldValue {
+                publishActionAllowed = hasContent && !isBeingPublished
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: Still thinking if this snippet deserves a method its own method, or not (refreshActionAllowedState). It's just a line... perhaps i'm overthinking this one!

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    func updated(isBeingPublished: Bool) {
+        self.isBeingPublished = isBeingPublished
+    }
+
+    var action: PostEditorAction {
+        return editorState.action()
+    }
+
+    var publishButtonText: String {
+        return editorState.action().publishActionLabel
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: `action.publishActionLabel`

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    }
+
+    var publishVerbText: String {
+        return editorState.action().publishingActionLabel
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: `action.publishingActionLabel` (nipticky, sorry!

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+    }
+
+    var isPostPostShown: Bool {
+        return editorState.action().isPostPostShown
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: `action.isPostPostShown` (sorry, nipticky++)

File: WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift (hunk 2 above)
Comment (astralbodies) [inline WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift]: @jleandroperez `didSet` isn't called when an `init` method sets the properties before `super` is called. Once super is called then didSet is called every future time the property is set. Stabby!!

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift (hunk 6 above)
Comment (astralbodies) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: Agreed - I'll turn action() into a protocol property.

**Comment by jleandroperez** [comment]
Drafts are displaying the `Publish` Action (vs `Update` in the Hybrid editor):

1. In a testing blog, create a Draft
2. Enable Aztec
3. Edit the Post. The top right action will read `Publish`

Is this expected?

**Comment by jleandroperez** [comment]
Glitch 2:

We've got an overlap between navigationBar Action Button, and the Aztec Title. We should either cap the Blog Title's Width, or indicate somehow else that we're in Aztec. (I'd vote for both perhaps!)

simulator screen shot feb 1 2017 11 57 51

**Comment by jleandroperez** [comment]
### Note 3:
The `post`'s setter observer is not refreshing the `PostEditorStateContext`. I suspect we should be reinstantiating the `PostEditorStateContext`'s instance, whenever the `Post` changes.

IE: From Revision to Original.

**Comment by jleandroperez** [comment]
### Update on Note 1:
Problem there is, precisely, that the `didChangeAction` delegate (which is in charge of setting the Publish Button's Title) is not getting called, initially.

We could either mimic KVO's "Post Initial Value" flag, or just add that to the `refreshInterface` helper?

**Comment by frosty** [comment]
A few notes from putting this through its paces:

* When writing a new post, the Publish button is only enabled when text is added to the _body_ of the post, not the title. On the hybrid editor, typing characters into the title field enables the button. We should probably do the same in Aztec? (you could save a post with only a title)
* If I set a new post to Scheduled (setting a publish date), the button changes to Schedule (as expected). If I change the post back to Published (by tapping Publish Immediately), the button title stays as Schedule. In fact, even if I change the post status to Draft the title is still Schedule. Not the case with the hybrid editor.
* I'm seeing the same as @jleandroperez – existing posts are all showing Publish for me. I tested with existing drafts, scheduled posts, and published posts. The button says Publish for all of them.
* The existing button says "Post" not "Publish". Are we changing this?

Taking a look through the code now... :)

File: WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift (hunk 1 above)
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift]: Could be a good idea – defaulting to `.draft`. Could also be used in `observeValue(forKey:...)`

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+        case .schedule:
+            return NSLocalizedString("Scheduling...", comment: "Text displayed in HUD while a post is being scheduled to be published.")
+        case .submitForReview:
+            return NSLocalizedString("Submitting for Review...", comment: "Text displayed in HUD while a post is being submitted for review.")
+        case .update:
+            return NSLocalizedString("Updating...", comment: "Text displayed in HUD while a draft or scheduled post is being updated.")
+        }
+    }
+
+    var isPostPostShown: Bool {
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: Looks like this isn't used yet and I might be misunderstanding its use, but with the hybrid editor we also show post-post for scheduled posts.

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+        }
+    }
+
+    fileprivate var originalPostStatus: PostStatus
+    fileprivate var userCanPublish: Bool
+    private var delegate: PostEditorStateContextDelegate?
+
+    fileprivate var hasContent = false {
+        didSet {
+            publishActionAllowed = hasContent && !isBeingPublished
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: I know it's only one line, but we could extract this to a private `updatePublishActionAllowed` func and call it from both `didSet`s?

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+    }
+}
+
+fileprivate class PostEditorStateSchedule: PostEditorActionState {
+    var action: PostEditorAction {
+        return .schedule
+    }
+
+    func updated(postStatus: PostStatus, context: PostEditorStateContext) -> PostEditorActionState {
+        return self
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: Presumably this is why I can't seem to get the button out of Schedule state once I've entered it?

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+        }
+
+        return self
+    }
+}
+
+fileprivate extension PostEditorActionState {
+    func isFutureDated(_ date: Date) -> Bool {
+        let oneMinute: TimeInterval = 60.0
+        let dateOneMinuteFromNow = Date(timeInterval: oneMinute, since: Date())
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: I would instead use `Calendar.current.date(byAdding: .minute, value: 1, to: date)`

Or you could potentially use `Calendar.current.compare(Date(), to: date, toGranularity: .minute)` to compare them instead.

File: WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift [hunk 11]
```diff
@@ -489,6 +527,43 @@ private extension AztecPostViewController {
...
+            let dateCreated = post.dateCreated ?? Date()
+            postEditorStateContext.updated(publishDate: dateCreated)
+        } else if keyPath == #keyPath(AbstractPost.content) {
+            let characterCount = post.content?.characters.count ?? 0
+            postEditorStateContext.updated(hasContent: characterCount > 0)
+        }
+    }
+
+    internal func context(_ context: PostEditorStateContext, didChangeAction: PostEditorAction) {
+        publishButton.title = context.publishButtonText
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Post/AztecPostViewController.swift]: In both of these delegate methods, is there a reason you're going via the context instead of using the passed in action / bool?

File: WordPress/WordPressTest/PostEditorStateTests.swift [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    func testContextNoContentPublishButtonDisabled() {
+        context = PostEditorStateContext(originalPostStatus: .publish, userCanPublish: true, delegate: self)
+
+        context.updated(hasContent: false)
+
+        XCTAssertFalse(context.isPublishButtonEnabled)
+    }
+
+    func testContextChangedNewPostToDraft() {
+        context = PostEditorStateContext(originalPostStatus: .draft, userCanPublish: true, delegate: self)
```
Comment (astralbodies) [inline WordPress/WordPressTest/PostEditorStateTests.swift]: I fixed this with a check-in - new posts will pass in a nil originalPostStatus now since it's not on the server. I was trying to match logic in Calypso which makes every post a draft when creating it which is something we should eventually emulate. 😄

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift (hunk 7 above)
Comment (astralbodies) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: It's not currently used yet - will be in another PR.

File: WordPress/Classes/ViewRelated/Post/PostEditorState.swift (hunk 8 above)
Comment (astralbodies) [inline WordPress/Classes/ViewRelated/Post/PostEditorState.swift]: Yup I'll make that change. 👍

File: WordPress/WordPressTest/PostEditorStateTests.swift (hunk 12 above)
Comment (frosty) [inline WordPress/WordPressTest/PostEditorStateTests.swift]: I definitely agree we should start every post off as a draft!

File: WordPress/WordPressTest/PostEditorStateTests.swift [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        XCTAssertEqual(PostEditorAction.publish, context.action, "should return Publish if the post is a draft")
+    }
+
+    func testContextUserCannotPublish() {
+        context = PostEditorStateContext(originalPostStatus: .draft, userCanPublish: false, delegate: self)
+
+        XCTAssertEqual(PostEditorAction.submitForReview, context.action, "should return 'Submit for Review' if the post is a draft and user can't publish")
+    }
+
+
```
Comment (astralbodies) [inline WordPress/WordPressTest/PostEditorStateTests.swift]: Added unit tests to cover these scenarios.

**Comment by astralbodies** [comment]
Responding to @frosty's main thread comments:

A few notes from putting this through its paces:

> When writing a new post, the Publish button is only enabled when text is added to the body of the post, not the title. On the hybrid editor, typing characters into the title field enables the button. We should probably do the same in Aztec? (you could save a post with only a title)

Done. Had to add a target/action for the editing changed event on the textview to watch for this. The postTitle property on the post isn't updated until editing is done.

> If I set a new post to Scheduled (setting a publish date), the button changes to Schedule (as expected). If I change the post back to Published (by tapping Publish Immediately), the button title stays as Schedule. In fact, even if I change the post status to Draft the title is still Schedule. Not the case with the hybrid editor.

Fixed.

> I'm seeing the same as @jleandroperez – existing posts are all showing Publish for me. I tested with existing drafts, scheduled posts, and published posts. The button says Publish for all of them.

Fixed.

> The existing button says "Post" not "Publish". Are we changing this?

I'm forcing us to match what Calypso says. I think @mattmiklic agreed when I asked about it. 😄

**Comment by astralbodies** [comment]
@jleandroperez Here are responses to your notes:

> Note 1:
> Drafts are displaying the Publish Action (vs Update in the Hybrid editor):

Fixed.

> Note 2:
> We've got an overlap between navigationBar Action Button, and the Aztec Title. We should either cap the Blog Title's Width, or indicate somehow else that we're in Aztec. (I'd vote for both perhaps!)

We talked about this and removing the navbar title makes the most sense. It's getting in the way and we'll mark the editor as being Aztec is some other way like slightly changing the title bar color or something. Will address in a different PR.

> Note 3:
> The post's setter observer is not refreshing the PostEditorStateContext. I suspect we should be re-instantiating the PostEditorStateContext's instance, whenever the Post changes.

For now I moved post revision creation to happen before the UI is set up. I think it makes sense we do the revision operation first.

> Note 4:
> Post's Status Context didChangeAction delegate method (which is in charge of setting the Publish Button's Title) is not getting called with the initial value.

I think letting the context emit the current state is a good idea. (There's a similar feature in RxSwift ... just sayin...) For now I'll leave it as there are a few things I want to add to the context (like the secondary publish button and post-post functionality) and I'll add this in.

**Comment by astralbodies** [comment]
@jleandroperez @frosty I *think* this is ready for another pass.

**Comment by mattmiklic** [comment]
> I'm forcing us to match what Calypso says. I think @mattmiklic agreed when I asked about it. 😄

💯. `Post` in the iOS editor is an outlier; both Calypso and wp-android both use `Publish`.

**Comment by frosty** [comment]
@astralbodies Changes look good! 👍 for all the comments you've added. Fixes the issues I had, and the tests all pass.

:shipit: from me, after the conflicts are fixed!

**Comment by jleandroperez** [comment]
Beautiful code @astralbodies, all looking good!!

:shipit: !!!

**Comment by astralbodies** [comment]
Thank you both so much for the time to review!
